app/step_planners/step_planner/move_barcodes_from_tube_rack_to_plate.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+module StepPlanner
+  # # This rule applies the barcodes from a rack into a plate
+  # {
+  #   ?plate se:a Plate .
+  #   ?rack se:a TubeRack .
+  #   ?rack se:mergeInto ?plate .
+  #   ?rack se:contains ?tube .
+  #   ?plate se:contains ?well .
+  #   ?tube se:location ?location .
+  #   ?tube se:barcode ?barcode .
+  #   ?well se:location ?location .
+  #   } => {
+  #     :step :addFacts { ?well se:barcode ?barcode . } .
+  #     :step :removeFacts { ?tube se:barcode ?barcode . } .
+  #     :step :addFacts { ?tube se:barcode  . } .
+  #     :step :addFacts { ?tube se:previousBarcode ?barcode . } .
+  #     :step :addFacts { ?tube se:appliedBarcodeTo ?well . } .
+  #     :step :addFacts { ?rack se:mergedInto ?plate .}.
+  #     :step :addFacts { ?plate se:mergedFrom ?rack .}.
+  #   }.
+  class MoveBarcodesFromTubeRackToPlate
+    attr_reader :asset_group
+
+    delegate :assets, to: :asset_group
+
+    def initialize(asset_group_id, _step_id)
+      @asset_group = AssetGroup.includes(assets: { facts: { object_asset: :facts } }).find(asset_group_id)
+    end
+
+    def assets_compatible_with_step_type
+      # @note I'm not sure why any? is accepted here, I'd have thought we'd want both
+      # plus failing silently in the event we have nothing to do seem undesireable
+      [plate, tube_rack].any?
+    end
+
+    def plate
+      @plate ||= assets.detect { |a| a.predicate_matching?('a', 'Plate') }
+    end
+
+    def tube_rack
+      @tube_rack ||= assets.detect { |a| a.predicate_matching?('a', 'TubeRack') }
+    end
+
+    def wells_for(asset)
+      asset.facts.with_predicate('contains').map(&:object_asset)
+    end
+
+    def index_wells_in(asset)
+      wells_for(asset).index_by do |w|
+        location_facts = w.facts.with_predicate('location').map(&:object).uniq
+        # We don't seem to have any case of this occurring in the production database, although there are some historic
```

Comment:
Perviously this just ignored invalid wells, but instead I've decided to would make more sense to fail.Its mostly theoretical, as all existing data appears valid.